cirq-google/cirq_google/circuit_serializer.py

```diff
@@ -0,0 +1,657 @@
+    into the constants table.  Usage is by passing a `cirq.Circuit`
+    to the `serialize()` method of the class, which will produce a
+    `Program` proto.  Likewise, the `deserialize` method will produce
+    a `cirq.Circuit` object from a `Program` proto.
+
+    This class is more performant than the previous `SerializableGateSet`
+    at the cost of some extendability.
+    """
+
+    def __init__(
+        self,
+        gate_set_name: str,
+    ):
+        """Construct the gate set.
+
+        Args:
+            gate_set_name: The name used to identify the gate set.
+            serializers: The OpSerializers to use for serialization.
+                Multiple serializers for a given gate type are allowed and
+                will be checked for a given type in the order specified here.
+                This allows for a given gate type to be serialized into
+                different serialized form depending on the parameters of the
+                gate.
+            deserializers: The OpDeserializers to convert serialized
+                forms of gates or circuits into Operations.
+        """
+        self.gate_set_name = gate_set_name
+
+    def serialize(
+        self,
+        program: cirq.Circuit,
+        msg: Optional[v2.program_pb2.Program] = None,
+        *,
+        arg_function_language: Optional[str] = None,
+        use_constants: bool = True,
+    ) -> v2.program_pb2.Program:
+        """Serialize a Circuit to cirq_google.api.v2.Program proto.
+
+        Args:
+            program: The Circuit to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            use_constants: Whether to use constants in serialization. This is
+                required to be True for serializing CircuitOperations.
+        """
+        if msg is None:
+            msg = v2.program_pb2.Program()
+        msg.language.gate_set = self.gate_set_name
+        if isinstance(program, cirq.Circuit):
+            constants: Optional[List[v2.program_pb2.Constant]] = [] if use_constants else None
+            raw_constants: Optional[Dict[Any, int]] = {} if use_constants else None
+            self._serialize_circuit(
+                program,
+                msg.circuit,
+                arg_function_language=arg_function_language,
+                constants=constants,
+                raw_constants=raw_constants,
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
+            if arg_function_language is None:
+                arg_function_language = arg_func_langs._infer_function_language_from_circuit(
+                    msg.circuit
+                )
+        else:
+            raise NotImplementedError(f'Unrecognized program type: {type(program)}')
+        msg.language.arg_function_language = arg_function_language
```

Comment:
nit: I'd suggest putting this up by gate set at the top of the function.

Reply:
Done.  Also got rid of the language inference.  There's no real need to iterate over the whole circuit again when we can just choose 'exp' and be done with it.